File: app/src/main/java/com/digis/android/task/data/source/remote/RemoteDataSource.java
package com.digis.android.task.data.source.remote;

import java.io.IOException;

import androidx.annotation.NonNull;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public final class RemoteDataSource<V> {
    private Call<V> mCall;
    private RemoteCallback<V> mCallback;

    public RemoteDataSource(@NonNull Call<V> call) {
        mCall = call;
    }

    public void execute(@NonNull RemoteCallback<V> callback) {
        mCallback = callback;
        if (!mCall.isExecuted()) {
            mCall.enqueue(new ResponseCallback());
        }
    }

    @SuppressWarnings("unused")
    public void retry(@NonNull RemoteCallback<V> callback) {
        mCallback = callback;
        mCall = mCall.clone();
        mCall.enqueue(new ResponseCallback());
    }

    public void cancel() {
        if (mCall != null && mCall.isExecuted()) {
            mCall.cancel();
        }
        mCallback.onCanceled();
    }

    private final class ResponseCallback implements Callback<V> {
        @Override
        public void onResponse(Call<V> call, Response<V> response) {
            V value = null;
            String errorBody = null;
            if (response.isSuccessful()) {
                value = response.body();
            }
            else {
                try { errorBody = response.errorBody().string(); }
                catch (IOException ignore) { }
            }
            HttpResponse<V> httpResponse = new HttpResponse<>(
                    response.code(), response.headers(), value, errorBody
            );
            mCallback.onResponse(httpResponse);
        }

        @Override
        public void onFailure(Call<V> call, Throwable throwable) {
            if (!call.isCanceled()) {
                mCallback.onFailure(throwable);
            }
        }
    }
}

File: app/src/main/java/com/digis/android/task/implementation/kotlin/data/source/remote/RemoteDataSource.kt
package com.digis.android.task.implementation.kotlin.data.source.remote

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException

class RemoteDataSource<V>(private var call: Call<V>) {
    private var callback: RemoteCallback<V>? = null

    fun execute(callback: RemoteCallback<V>) {
        this.callback = callback
        if (!call.isExecuted) {
            call.enqueue(ResponseCallback())
        }
    }

    @Suppress("unused")
    fun retry(callback: RemoteCallback<V>) {
        this.callback = callback
        call = call.clone()
        call.enqueue(ResponseCallback())
    }

    fun cancel() {
        if (call.isExecuted) {
            call.cancel()
        }
        callback?.onCanceled()
    }

    private inner class ResponseCallback : Callback<V> {
        override fun onResponse(call: Call<V>?, r: Response<V>?) {
            var value: V? = null
            var errorBody: String? = null
            val response = r!!
            if (response.isSuccessful) {
                value = response.body()
            } else {
                try { errorBody = response.errorBody().string() }
                catch (ignore: IOException) { }
            }
            val httpResponse = HttpResponse(
                    response.code(), response.headers(), value, errorBody
            )
            callback?.onResponse(httpResponse)
        }

        override fun onFailure(call: Call<V>?, t: Throwable?) {
            val isCanceled = call?.isCanceled ?: false
            if (!isCanceled) callback?.onFailure(t!!)
        }
    }
}